src/Couchbase.Lite.Shared/API/Database/Database.cs

```diff
@@ -816,10 +816,79 @@ public void RemoveChangeListener(ListenerToken token)
         public bool Save([NotNull]MutableDocument document, ConcurrencyControl concurrencyControl)
         {
             var doc = CBDebug.MustNotBeNull(WriteLog.To.Database, Tag, nameof(document), document);
-            return Save(doc, concurrencyControl, false);
+            return Save(doc, null, concurrencyControl, false);
         }
 
-        #if CBL_LINQ
+        /// <summary>
+        /// Saves a document to the database. When write operations are executed concurrently, 
+        /// and if conflicts occur, conflict handler will be called. Use the handler to directly
+        /// edit the document.Returning true, will save the document. Returning false, will cancel
+        /// the save operation.
+        /// </summary>
+        /// <param name="document">The document to save</param>
+        /// <param name="conflictHandler">The conflict handler block which can be used to resolve it.</param> 
+        /// <returns><c>true</c> if the save succeeded, <c>false</c> if there was a conflict</returns>
+        public bool Save([NotNull]MutableDocument document, Func<MutableDocument, Document, bool> conflictHandler)
+        {
+            var doc = CBDebug.MustNotBeNull(WriteLog.To.Database, Tag, nameof(document), document);
+            Debug.Assert(conflictHandler != null);
+            Document currDoc = null;
+
+            while (true) {
+                var success = Save(doc, currDoc, ConcurrencyControl.FailOnConflict, false);
+                if (success)
+                    return success;
+                var committed = false;
+                try {
+                    LiteCoreBridge.Check(err => Native.c4db_beginTransaction(_c4db, err));
```

Comment:
You cannot hold this transaction the whole time since that will block write access to the database during conflict resolution.  The attempted save should be inside of `DoLocked` and a transaction.  Then if the result is false, the conflict resolution needs to take place outside the lock and transaction, and then loop around again into another lock and transaction.